lorcon/lorcon.3
.\" $Id: $
.TH LORCON 3 "12 Sep 2009"
.SH NAME
LORCON \- 802.11 Wireless Packet Injection Library
.SH SYNOPSIS
.nf
.ft B
#include <lorcon2/lorcon.h>
#include <lorcon2/lorcon_packet.h>
.ft
.LP
.nf
.ft B
typedef void (*lorcon_handler)(lorcon_t *, lorcon_packet_t *, u_char *user);
.ft
.LP
.ft B
const char *lorcon_get_error(lorcon_t *context)
.ft
.LP
.ft B
lorcon_driver_t *lorcon_list_drivers()
lorcon_driver_t *lorcon_find_driver(const char *driver)
lorcon_driver_t *lorcon_auto_driver(const char *interface)
void lorcon_free_driver_list(lorcon_driver_t *listhead)
.ft
.LP
.ft B
lorcon_t *lorcon_create(const char *interface, lorcon_driver_t *driver)
voir lorcon_close(lorcon_t *context)
void lorcon_free(lorcon_t *context)
.ft
.LP
.ft B
int lorcon_open_inject(lorcon_t *context)
int lorcon_open_monitor(lorcon_t *context)
int lorcon_open_injmon(lorcon_t *context)
.ft
.LP
.ft B
void lorcon_set_vap(lorcon-t *context, const char *vap)
const char *lorcon_et_vap(lorcon_t *context)
const char *lorcon_get_capiface(lorcon_t *context)
const char *lorcon_get_driver_name(lorcon_t *context)
.ft
.LP
.ft B
int lorcon_get_datalink(lorcon_t *context)
int lorcon_set_datalink(lorcon_t *context, int dlt)
.ft
.LP
.ft B
int lorcon_get_channel(lorcon_t *context)
int lorcon_set_channel(lorcon_t *context, int channel)
.ft
.LP
.ft B
pcap_t *lorcon_get_pcap(lorcon_t *context)
int lorcon_get_selectable_fd(lorcon_t *context)
.ft
.LP
.ft B
int lorcon_next_ex(lorcon_t *context, lorcon_packet_t **packet)
.ft
.LP
.ft B
int lorcon_set_filter(lorcon_t *context, const char *filter)
int lorcon_loop(lorcon_t *context, int count, lorcon_handler callback, u_char *user)
int lorcon_dispatch(lorcon_t *context, int count, lorcon_handler callback, u_char *user)
void lorcon_breakloop(lorcon_t *context);
.ft
.LP
.ft B
int lorcon_send_bytes(lotcon_t *context, int length, u_char *bytes)
int lorcon_inject(lorcon_t *context, lorcon_packet_t *packet)
.ft
.ft B
lorcon_packet_t *lorcon_packet_from_pcap(struct lorcon *context, const struct pcap_pkthdr *h, const u_char *bytes)
lorcon_packet_t *lorcon_packet_from_dot3(u_char *bssid, int dot11_direction, u_char *data, int length)
.ft
.fi

.SH DESCRIPTION

LORCON - Loss Of Radio CONnectivity
.LP
The LORCON packet library provides a higher-level interface to transmit and
process wireless packets.  

.PP
.SH AUTHORS
.LP
Joshua Wright <jwright@hasborg.com>
.LP
Mike Kershaw <dragorn@kismetwireless.net>
.LP

.SH BUGS
Please send problems, bugs, questions, desirable enhancements, patches and source code contributions to:
.LP
.RS
lorcon@802.11ninja.net

.SH RESOURCES
Please see http://802.11ninja.net/lorcon for updates to LORCON, documentation and the bug database.
.RE


mdk3/CHANGELOG
MDK3 Changelog

version 6
* Amok Mode now works on Ad-Hoc and MANET networks (WARNING: Clients may not reconnect automatically,
  so they may stay disconnected after the attack stopped!)
* Removed duplicate WPA downgrade in Deauth Mode (sry for the confusion)
* SSID Bruteforce Mode understands 0 and 1 byte placeholders in hidden SSIDs now
* GCC 4.4 support, all warnings and extra warnings fixed
* Whitelists and Blacklists in Amok Mode are re-read periodically every 3 seconds.
  You can use this to dynamically allow or block hosts with scripts.
* A lot of small bugfixes

version 5
* Enhanced MAC-Filter Bruteforce Mode
* Another WDS/WIDS/WIPS Confusion Test
* Amok Mode supports QoS packets
* Michael Countermeasure Exploit (also known as TKIP QoS Exploit)
  can shut down APs using TKIP encryption and QoS Extension with 1 sniffed and 2 injected QoS Data Packets.
* WPA-Downgrade Test - deauthenticates Stations and APs sending WPA encrypted packets.
  With this test you can check if the sysadmin will try setting his network to WEP or disable encryption.
  mdk3 will let WEP and unencrypted clients work, so if the sysadmin simply thinks "WPA is broken" he sure isn't the right one for this job.
  (this can/should be combined with social engineering)

version 4
* Added experimental high speed MAC-Filter Bruteforce Mode

version 3
* Added a channel hopper for Amok Mode
* Added WIDS confusion mode
* fresh & bugfixed osdep included
* Fixed White- and Blacklisting again...

version 2
* More Documentation
* Added -Wall to Makefile to always keep the code clean
* Fixed the Warnings produced by -Wall ;)
* Updated osdep
* Merged some patches from Andy Green to clean up the code
* Poured some holy water all over the code, because mdk3 is used by professionals
* Added intelligent Authentication DoS mode
* Fixed White- and Blacklist function in Amok mode

version 1
* NOW USING OSDEP INJECTION from aircrack-ng project
  -> mdk2 should now run on LINUX and FREEBSD (and soon more)
* Started writing some docs (now that mdk3 may soon run on Windows, the kids may need it...)


MDK2 version history:
version 37 (has never been released, went directly to mdk3)
* Better Madwifi-ng handling
* Blacklist mode fixed

version 36
* Blacklist support for Deauth Mode
* Fixed some minor gcc warnings
* Advanced FakeAP mode. Reads SSIDs and MACs from file.
  See example file.

version 35
* telek0miker seems to have his EAPOL-Start test fixed
* Help screen is now splitted into several sexy parts.

version 34
* Compilation fix for PPC architecture
* Le_Vert added all copyright and license informations. (thx)
* Rewrite Makefile.

version 33
* Added telek0miker's EAPOL-Logoff Test (thanks again)

version 32
* Added telek0miker's EAPOL-Start flood (thanx!)

version 31
* fixed race condition in bruteforce mode

version 30
AKA "Hell, I don't know how to do the versioning!"
* Just counting version numbers now, no beta and dev versions
* Added beacon flood variety patch from ducttape (THANKS!)
* Ace's SSID Bruteforcer works now

beta version 3
AKA "Zuo Meng Hua, I love you!"-Edition
* Used thefkboss' "motherfucker" version
* Added Whitelist support for deauth mode
* Updated TODO list, looks good to have more "DONE" in there ;)
* Cleaned up a little
* Stabilized and cleaned SSID Bruteforcer
* Added Ace's SSID Bruteforcer - HUGE THANKS! (Doesnt work yet...)
* Added Hirte's TKIP Shutdown Exploit - THANK YOU!
* Makefile fixed

development version 0028
AKA BETA 2
* cleaned up code, removed gcc warnings
* fixed -e/-s confusion in probe mode
* fixed signedness bug in beacon mode display
* now using Makefile from Zero_Chaos (thx!)
* implemented Ace's SSID Bruteforce (thx!)
* probe checks works on rt2570 using my 1.4.0 driver (required for SSID Bruteforce)

development version 0027
* speeding up tests which need to sniff
* Added option to use valid MAC adresses from OUI database

development version 0026
* Added SSID wordlist mode (thanks to moongray)
* Added SSID lenght checkings and warnings
* Total rewrite of Deauth-Amok-Mode, now it's even more aggressive ;)

development version 0024
* fixed header files for newer distributions (SuSE 10.1, FC5, maybe others)
* fixed very stupid bug causing segfault in some cases
* Added some more options to Fake AP mode (thanks to Antragon)

beta version 1 - had the feeling to make one... ;)
* added fully automatic deauthentication/disassociation mode (Amok-Mode!)

development version 00xx
* added probe mode
* probe check does not work on rt2570, need to investigate why

development version 0019
* tweaked auth dos to react faster in conditions with low traffic
* additional information and out of range warning in auth DoS checker
* Beacon Flood mode now supports channel hopping

development version 0016
* auth dos mode more fixup
* auth dos checker is now kind of mature (perhaps needs tweaking)
* cleaning up code, removing unused variables
* adding a lot of comments to make easier to understand

development version 0014
* auth dos mode fixed
* auth dos checker working very unreliable
* New method to generate random SSIDs

development version 0012
* basic beacon mode functional
* very basic and buggy auth dos mode without check for success
* this code was mostly copied from old mdk version

preparing version 0000
* preparing copied functions from aireplay


pylorcon2/lorcon.py
#!/usr/bin/env python
'''
pylorcon2 - Python wrapper for LORCON 2
By: Tom Wambold <tom5760@gmail.com>
Home: http://github.com/tom5760/pylorcon2/
LORCON: http://802.11ninja.net/

'''

import ctypes
from ctypes.util import find_library

class LorconError(Exception): pass

class StructLorcon(ctypes.Structure): pass

# Function Pointers
LORCON_DRV_INIT = ctypes.CFUNCTYPE(ctypes.c_int, ctypes.POINTER(StructLorcon))
LORCON_DRV_PROBE = ctypes.CFUNCTYPE(ctypes.c_int, ctypes.c_char_p)

# Forward declare this struct
class StructDriver(ctypes.Structure):
    pass

# We need to do this since this struct references itself
StructDriver._fields_ = (
        ('name', ctypes.c_char_p),
        ('details', ctypes.c_char_p),

        ('init_func', LORCON_DRV_INIT),
        ('probe_func', LORCON_DRV_PROBE),

        ('next', ctypes.POINTER(StructDriver)),
    )

class Lorcon(object):
    'Represents the LORCON library.'

    def __init__(self, lib_path=None):
        '''Instantiate the LORCON library.

        lib_path - Path to LORCON shared library.  If None, it is autodetected.

        '''

        if lib_path is None:
            # Look for -lorcon2
            lib_path = find_library('orcon2')

        self.lib = ctypes.cdll.LoadLibrary(lib_path)

        # Set argument and return types for functions
        self.lib.lorcon_list_drivers.argtypes = None
        self.lib.lorcon_list_drivers.restype = ctypes.POINTER(StructDriver)

        self.lib.lorcon_find_driver.argtypes = [ctypes.c_char_p]
        self.lib.lorcon_find_driver.restype = ctypes.POINTER(StructDriver)

        self.lib.lorcon_auto_driver.argtypes = [ctypes.c_char_p]
        self.lib.lorcon_auto_driver.restype = ctypes.POINTER(StructDriver)

        self.lib.lorcon_free_driver_list.argtypes = [ctypes.POINTER(StructDriver)]
        self.lib.lorcon_free_driver_list.restype = None

        self.lib.lorcon_create.argtypes = [ctypes.c_char_p,
                ctypes.POINTER(StructDriver)]
        self.lib.lorcon_create.restype = ctypes.POINTER(StructLorcon)

        self.lib.lorcon_free.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_free.restype = None

        self.lib.lorcon_get_error.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_get_error.restype = ctypes.c_char_p

        self.lib.lorcon_get_timeout.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_get_timeout.restype = ctypes.c_int

        self.lib.lorcon_set_timeout.argtypes = [ctypes.POINTER(StructLorcon),
                ctypes.c_int]
        self.lib.lorcon_set_timeout.restype = None

        self.lib.lorcon_set_timeout.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_set_timeout.restype = ctypes.c_int

        self.lib.lorcon_open_inject.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_open_inject.restype = ctypes.c_int
        self.lib.lorcon_open_inject.errcheck = self._errcheck_int

        self.lib.lorcon_open_monitor.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_open_monitor.restype = ctypes.c_int
        self.lib.lorcon_open_monitor.errcheck = self._errcheck_int

        self.lib.lorcon_get_vap.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_get_vap.restype = ctypes.c_char_p

        self.lib.lorcon_set_vap.argtypes = [ctypes.POINTER(StructLorcon),
                ctypes.c_char_p]
        self.lib.lorcon_set_vap.restype = None

        self.lib.lorcon_get_capiface.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_get_capiface.restype = ctypes.c_char_p

        self.lib.lorcon_get_driver_name.argtypes = [
                ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_get_driver_name.restype = ctypes.c_char_p

        # TODO: These are not implemented in Lorcon2 yet
        #self.lib.lorcon_get_datalink.argtypes = [ctypes.POINTER(StructLorcon)]
        #self.lib.lorcon_get_datalink.restype = ctypes.c_int

        #self.lib.lorcon_set_datalink.argtypes = [ctypes.POINTER(StructLorcon),
        #        ctypes.c_int]
        #self.lib.lorcon_set_datalink.restype = ctypes.c_int
        #self.lib.lorcon_set_datalink.errcheck = self._errcheck_int

        self.lib.lorcon_get_channel.argtypes = [ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_get_channel.restype = ctypes.c_int

        self.lib.lorcon_set_channel.argtypes = [ctypes.POINTER(StructLorcon),
                ctypes.c_int]
        self.lib.lorcon_set_channel.restype = ctypes.c_int
        self.lib.lorcon_set_channel.errcheck = self._errcheck_int

        self.lib.lorcon_get_hwmac.argtypes = [ctypes.POINTER(StructLorcon),
                ctypes.POINTER(ctypes.POINTER(ctypes.c_uint8))]
        self.lib.lorcon_get_hwmac.restype = ctypes.c_int
        self.lib.lorcon_get_hwmac.errcheck = self._errcheck_int

        self.lib.lorcon_set_hwmac.argtypes = [ctypes.POINTER(StructLorcon),
                ctypes.c_int, ctypes.POINTER(ctypes.c_uint8)]
        self.lib.lorcon_set_hwmac.restype = ctypes.c_int
        self.lib.lorcon_set_hwmac.errcheck = self._errcheck_int

        self.lib.lorcon_get_pcap.argtypes = [ctypes.POINTER(StructLorcon)]
        # TODO: Maybe make a real ctypes type for this?  Should be struct pcap_t
        self.lib.lorcon_get_pcap.restype = ctypes.c_void_p

        self.lib.lorcon_get_selectable_fd.argtypes = [
                ctypes.POINTER(StructLorcon)]
        self.lib.lorcon_get_selectable_fd.restype = ctypes.c_int

        # TODO: lorcon_next_ex goes here

        self.lib.lorcon_set_filter.argtypes = [ctypes.POINTER(StructLorcon),
                ctypes.c_char_p]
        self.lib.lorcon_set_filter.restype = ctypes.c_int
        self.lib.lorcon_set_filter.errcheck = self._errcheck_int

        # TODO: lorcon_set_compiled_filter goes here
        # TODO: lorcon_pcap_loop goes here
        # TODO: lorcon_pcap_dispatch goes here

        self.lib.lorcon_breakloop.argtypes = None
        self.lib.lorcon_breakloop.restype = None

        # TODO: Wrap the lorcon_packet stuff
        #self.lib.lorcon_inject.argtypes = [ctypes.POINTER(StructLorcon),
        #        ctypes.POINTER(StructPacket)]
        #self.lib.lorcon_inject.restype = ctypes.c_int
        #self.lib.lorcon_inject.errcheck = self._errcheck_int

        self.lib.lorcon_send_bytes.argtypes = [ctypes.POINTER(StructLorcon),
                ctypes.c_int, ctypes.POINTER(ctypes.c_ubyte)]
        self.lib.lorcon_send_bytes.restype = ctypes.c_int
        self.lib.lorcon_send_bytes.errcheck = self._errcheck_int

        self.lib.lorcon_get_version.argtypes = None
        self.lib.lorcon_get_version.restype = ctypes.c_ulong

        self.lib.lorcon_add_wepkey.argtypes = [ctypes.POINTER(StructLorcon),
                ctypes.POINTER(ctypes.c_ubyte), ctypes.POINTER(ctypes.c_ubyte)]
        self.lib.lorcon_add_wepkey.restype = ctypes.c_int
        self.lib.lorcon_add_wepkey.errcheck = self._errcheck_int

    def list_drivers(self):
        'List all drivers LORCON is compiled with.'
        return _generate_driver_list(self.lib, self.lib.lorcon_list_drivers())

    def find_driver(self, driver):
        return Driver(self.lib, self.lib.lorcon_find_driver(driver))

    def auto_driver(self, interface):
        return Driver(self.lib, self.lib.lorcon_auto_driver(interface))

    def create(self, interface, driver):
        return Context(self.lib, self.lib.lorcon_create(interface, driver))

    def _errcheck_int(self, result, func, args):
        'Raises an exception if the return value is less than 0'
        if result < 0:
            raise LorconError(self.lib.lorcon_get_error(args[0]))
        return result

class Driver(object):
    '''Represents a driver struct.

    Automatically frees memory when this object is garbage collected.

    '''
    def __init__(self, lib, ptr):
        '''lib = reference to the lorcon library
           ptr = ctypes pointer to a StructDriver

       '''
        self.lib = lib
        self._as_parameter_ = ptr
        self.contents = ptr.contents

    def __del__(self):
        self.lib.lorcon_free_driver_list(self)

    def __str__(self):
        return self.name

    name = property(lambda self: self.contents.name)
    details = property(lambda self: self.contents.details)

class Context(object):
    'Represents a LORCON context struct.'

    def __init__(self, lib, ptr):
        '''lib = reference to the lorcon library
           ptr = ctypes pointer to a StructLorcon

        '''
        self.lib = lib
        self._as_parameter_ = ptr
        self.contents = ptr.contents

    def __del__(self):
        self.lib.lorcon_free(self)

    def open_inject(self):
        'Open an interface for inject.'
        self.lib.lorcon_open_inject(self)

    def open_monitor(self):
        'Open an interface in monitor mode (may also enable injection).'
        self.lib.lorcon_open_monitor(self)

    def open_injmon(self):
        'Open an interface in inject+monitor mode.'
        self.lib.lorcon_open_injmon(self)

    def close(self):
        'Close interface.'
        self.lib.lorcon_close(self)

    def fileno(self):
        'For use in the "select" package'
        return self.selectable_fd

    def next_ex(self):
        raise NotImplementedError()

    def set_filter(self, filter):
        self.lib.lorcon_set_filter(self, filter)

    def set_compiled_filter(self, filter):
        raise NotImplementedError()

    def pcap_loop(self, count, callback):
        raise NotImplementedError()

    def pcap_dispatch(self, count, callback):
        raise NotImplementedError()

    def pcap_breakloop(self):
        self.lib.lorcon_breakloop(self)

    def inject(self, packet):
        # TODO: Wrap the lorcon_packet stuff
        raise NotImplementedError()
        #return self.lib.lorcon_inject(self, packet)

    def send_bytes(self, data):
        buf = (ctypes.c_ubyte * len(data))(*[ord(x) for x in data])
        return self.lib.lorcon_send_bytes(self, len(buf), buf)

    def add_wepkey(self, bssid, key):
        return self.lib.lorcon_add_wepkey(self, bssid, key, len(key))

    # Properties (Replaces getters/setters

    version = property(lambda self: self.lib.lorcon_get_version(self))

    error = property(lambda self: self.lib.lorcon_get_error(self),
            doc='Return the most recent error.')

    timeout = property(
            lambda self: self.lib.lorcon_get_timeout(self),
            lambda self, x: self.lib.lorcon_set_timeout(self, x),
            doc='''Set a capture timeout (equivalent to the timeout value in
            pcap_open, but with implications for non-pcap sources as well).
            Timeout value is in ms.''')

    vap = property(
            lambda self: self.lib.lorcon_get_vap(self),
            lambda self, x: self.lib.lorcon_set_vap(self, x),
            doc='Set/Get the VAP (if we use VAPs).')

    capiface = property(
            lambda self: self.lib.lorcon_get_capiface(self),
            doc='Get the interface we\'re capturing from.')

    driver_name = property(
            lambda self: self.lib.lorcon_get_driver_name(self),
            doc='Get the driver.')

    # TODO: This isn't defined in Lorcon2 yet
    #datalink = property(
    #        lambda self: self.lib.lorcon_get_datalink(self),
    #        lambda self, x: self.lib.lorcon_set_datalink(self, x),
    #        doc='Datalink layer info')

    channel = property(
            lambda self: self.lib.lorcon_get_channel(self),
            lambda self, x: self.lib.lorcon_set_channel(self, x),
            doc='Get/set channel/frequency')

    pcap = property(
            lambda self: self.lib.lorcon_get_pcap(self),
            doc='Get a pcap_t')

    selectable_fd = property(
            lambda self: self.lib.lorcon_get_selectable_fd(self))

    # These properties are a bit more complicated, so define them using
    # decorators

    @property
    def hwmac(self):
        '''Get/set MAC address, returns length of MAC and allocates in **mac,
        caller is responsible for freeing this memory.  Different PHY types may
        have different MAC lengths.

        For 802.11, MAC is always 6 bytes.

        A length of 0 indicates no set MAC on this PHY.  Negative numbers
        indicate error fetching MAC from hardware.

        '''
        mac = ctypes.POINTER(ctypes.c_uint8)()
        maclen = self.lib.lorcon_get_hwmac(self, ctypes.byref(mac))
        if maclen:
            return mac[0:maclen]
        raise LorconError('No MAC on this PHY')

    @hwmac.setter
    def hwmac(self, mac):
        'mac should be a list of 8 bit integers'
        mac_array = (ctypes.c_uint8 * len(mac))(*mac)
        return self.lib.lorcon_set_hwmac(self, len(mac_array), mac_array)

def _generate_driver_list(lib, ptr):
    '''Creates a list of Driver objects from a pointer to a driver list from
    lorcon_list_drivers.

    This makes sure each Driver object can be independently garbage collected.

    '''
    driver_list = []
    while True:
        try:
            driver_list.append(Driver(lib, ptr))
        except ValueError:
            break
        ptr = ptr.contents.next

    # Remove the "next" pointers in each driver, so that they can be garbage
    # collected independently
    for driver in driver_list:
        driver.contents.next = None

    return driver_list

if __name__ == '__main__':
    l = Lorcon('/tmp/lorcon/lib/liborcon2.so')
    driver = l.auto_driver('wlan0')
    context = l.create('wlan0', driver)
    context.open_inject()
    context.send_bytes('Hello World')
